### PR TITLE
[admin-tool] Enable venice-admin-tool to parse JSON map arguments from command line

### LINKED ISSUE #647: [BUG] venice-admin-tool doesn't parse map arguments correctly
### Willingness to contribute

No. I cannot contribute a bug fix at this time.

### Venice version

All versions

### System information

-- all platforms, this is a code bug --

### The Problem
In venice-admin-tool some commands are meant to take json maps as arguments.  Primary examples of this include using the --update-store command to set a custom partitioner or setting store views.  Unfortunately, right now the tool doesn't parse these well from command line (and it was heretofore undetected because java tests which invoke the api directly don't have this problem).

### Example Input
java -jar venice-admin-tool-all.jar --update-store --url http://venice-controller:1576 --cluster venice-cluster --store TestGpuJobsDataStoreV1 --storage-view-configs '{"changeCaptureView": {"viewClassName": "com.linkedin.venice.views.ChangeCaptureView","params": {}}}'

### Tracking information

_No response_

### Code to reproduce bug

```
zpolicze@zpolicze-mn2 libs % java -jar venice-admin-tool-all.jar --update-store --url http://ltx1-app6326.stg.linkedin.com:1576 --cluster mt-0 --store TestGpuJobsDataStoreV1 -storage-view-configs '{"changeCaptureView": {"viewClassName": "com.linkedin.venice.views.ChangeCaptureView","params": {}}}'
[WARN] Running admin tool without SSL.
{
  "error" : "storage-view-configs must be key value pairs separated by comma, but value: {\"changeCaptureView\": {\"viewClassName\": \"com.linkedin.venice.views.ChangeCaptureView\",\"params\": {}}}"
}
Exception in thread "main" com.linkedin.venice.exceptions.VeniceException: storage-view-configs must be key value pairs separated by comma, but value: {"changeCaptureView": {"viewClassName": "com.linkedin.venice.views.ChangeCaptureView","params": {}}}
	at com.linkedin.venice.utils.Utils.parseCommaSeparatedStringMapFromString(Utils.java:325)
	at com.linkedin.venice.AdminTool.lambda$stringMapParam$7(AdminTool.java:992)
	at com.linkedin.venice.AdminTool.genericParam(AdminTool.java:1010)
	at com.linkedin.venice.AdminTool.stringMapParam(AdminTool.java:992)
	at com.linkedin.venice.AdminTool.getUpdateStoreQueryParams(AdminTool.java:1090)
	at com.linkedin.venice.AdminTool.updateStore(AdminTool.java:1015)
	at com.linkedin.venice.AdminTool.main(AdminTool.java:318)
```

### What component(s) does this bug affect?

- [ ] `Controller`: This is the control-plane for Venice. Used to create/update/query stores and their metadata.
- [ ] `Router`: This is the stateless query-routing layer for serving read requests.
- [ ] `Server`: This is the component that persists all the store data.
- [ ] `VenicePushJob`: This is the component that pushes derived data from Hadoop to Venice backend.
- [ ] `VenicePulsarSink`: This is a Sink connector for Apache Pulsar that pushes data from Pulsar into Venice.
- [ ] `Thin Client`: This is a stateless client users use to query Venice Router for reading store data.
- [ ] `Fast Client`: This is a stateful client users use to query Venice Server for reading store data.
- [ ] `Da Vinci Client`: This is an embedded, stateful client that materializes store data locally.
- [ ] `Alpini`: This is the framework that fast-client and routers use to route requests to the storage nodes that have the data.
- [ ] `Samza`: This is the library users use to make nearline updates to store data.
- [X] `Admin Tool`: This is the stand-alone client used for ad-hoc operations on Venice.
- [ ] `Scripts`: These are the various ops scripts in the repo.

### PR DESCRIPTION
**Pull request by elijahgrimaldi**
## Enable venice-admin-tool to parse JSON map arguments from command line

Added correct JSON parsing for command line inputs to admin-tool to read maps as JSON

Resolves #647 

## How was this PR tested?
CI

## Does this PR introduce any user-facing changes?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure to explain your proposed changes and call out the behavior change.

### CODE REVIEW

File: internal/venice-common/src/main/java/com/linkedin/venice/utils/Utils.java [hunk 1]
```diff
@@ -313,11 +315,28 @@ public static Map<String, String> parseCommaSeparatedStringMapFromString(String
     try {
```
Comment (nisargthakkar) [inline internal/venice-common/src/main/java/com/linkedin/venice/utils/Utils.java:315]: Let's rename the function and update it's javadoc too, please.

File: internal/venice-common/src/main/java/com/linkedin/venice/utils/Utils.java [hunk 2]
```diff
@@ -313,11 +315,28 @@ public static Map<String, String> parseCommaSeparatedStringMapFromString(String
...
       Map<String, String> map = new HashMap<>();
       if (!value.isEmpty()) {
-        Arrays.stream(value.split(",")).map(s -> s.split("=")).forEach(strings -> map.put(strings[0], strings[1]));
+        ObjectMapper objectMapper = new ObjectMapper();
```
Comment (nisargthakkar) [inline internal/venice-common/src/main/java/com/linkedin/venice/utils/Utils.java:318]: Let's use `ObjectMapperFactory.getInstance()` instead of creating a new object

File: internal/venice-common/src/main/java/com/linkedin/venice/utils/Utils.java [hunk 3]
```diff
@@ -313,11 +315,28 @@ public static Map<String, String> parseCommaSeparatedStringMapFromString(String
...
+        JsonNode rootNode = objectMapper.readTree(value);
+
+        rootNode.fieldNames().forEachRemaining(key -> {
+          JsonNode childNode = rootNode.get(key);
+          if (childNode.isObject()) {
+            map.put(key, childNode.toString());
+          } else {
+            map.put(key, childNode.asText());
+          }
+        });
```
Comment (nisargthakkar) [inline internal/venice-common/src/main/java/com/linkedin/venice/utils/Utils.java:328]: Instead of parsing this JSON string manually, would it work if we use:
```Map<String, String> map = mapper.readValue(value, new TypeReference<Map<String, String>>() {});``` ?

File: internal/venice-common/src/main/java/com/linkedin/venice/utils/Utils.java [hunk 4]
```diff
@@ -313,11 +315,28 @@ public static Map<String, String> parseCommaSeparatedStringMapFromString(String
...
       }
       return map;
-    } catch (Exception e) {
-      throw new VeniceException(fieldName + " must be key value pairs separated by comma, but value: " + value);
+    } catch (IOException jsonException) {
+      try {
+        Map<String, String> map = new HashMap<>();
+        Arrays.stream(value.split(",")).map(s -> s.split("=")).forEach(strings -> map.put(strings[0], strings[1]));
```
Comment (nisargthakkar) [inline internal/venice-common/src/main/java/com/linkedin/venice/utils/Utils.java:334]: Let's not carry this tech-debt over. Since it is admin-tool, it is okay to make breaking changes for now since it is only used ad-hoc

File: internal/venice-common/src/test/java/com/linkedin/venice/utils/UtilsTest.java [hunk 5]
```diff
@@ -183,4 +183,21 @@ public void testParseMap() {
   public void testSanitizingStringForLogger() {
     Assert.assertEquals(Utils.getSanitizedStringForLogger(".abc.123."), "_abc_123_");
   }
+
+  @Test
+  public void testParseCommaSeparatedStringMapFromString() {
```
Comment (nisargthakkar) [inline internal/venice-common/src/test/java/com/linkedin/venice/utils/UtilsTest.java:188]: Let's rename this test along with the original function name since it is no longer a comma separated string map

**Review by nisargthakkar — CHANGES_REQUESTED** [review]
Thanks for the contribution

File: internal/venice-common/src/main/java/com/linkedin/venice/utils/Utils.java (hunk 4 above)
Comment (elijahgrimaldi) [reply internal/venice-common/src/main/java/com/linkedin/venice/utils/Utils.java:334]: Hey there, thanks for all the input I will make the changes you suggested. Can you explain a bit more about what you mean by tech-debt and how it applies? I have not heard this term before.

Thanks!

Comment (nisargthakkar) [reply internal/venice-common/src/main/java/com/linkedin/venice/utils/Utils.java:334]: Of course. Tech-debt is a term used to refer to anything that eventually becomes a maintenance overhead. In long-running software development, there are many iterations to a codebase and the behavior of the system evolves.

Similar to this PR, say you want to replace an existing functionality with another one. One way is to However, we've now accumulated users that are using the previous functionality of product and new features should not break the functionality that old users are using. So, we generally support both modes for some time till we can gracefully deprecate the old mode.

But in this PRs case, the code we're touching is the admin tool which is a tool we use to do ad-hoc maintenance operations against a running Venice cluster. The users of these tools are expected to be the operators of the Venice cluster and it is not widely used by users of Venice. So, in some cases like this, it is okay to break the backward compatibility in favor of reducing tech-debt.

Does this help? Also, I find Reddit ELI5's as a good resource for understanding any new concept. Here's the one for Tech-debt: https://www.reddit.com/r/explainlikeimfive/comments/5oxfe9/eli5_technical_debt/

Comment (elijahgrimaldi) [reply internal/venice-common/src/main/java/com/linkedin/venice/utils/Utils.java]: I appreciate the explanation, I just made some changes so let me know what you think and if anything needs to be improved still.

File: internal/venice-common/src/main/java/com/linkedin/venice/utils/Utils.java [hunk 6]
```diff
@@ -306,18 +308,21 @@ public static boolean parseBooleanFromString(String value, String fieldName) {
...
+        ObjectMapper objectMapper = ObjectMapperFactory.getInstance();
+        return objectMapper.readValue(value, new TypeReference<Map<String, String>>() {
+        });
       }
       return map;
-    } catch (Exception e) {
-      throw new VeniceException(fieldName + " must be key value pairs separated by comma, but value: " + value);
+    } catch (IOException jsonException) {
+      throw new VeniceException(
+          fieldName + " must be a valid JSON object or key-value pairs separated by comma, but value: " + value);
```
Comment (nisargthakkar) [inline internal/venice-common/src/main/java/com/linkedin/venice/utils/Utils.java:325]: Let's edit this error message now that we don't support comma-separated key-value pairs

File: internal/venice-common/src/test/java/com/linkedin/venice/utils/UtilsTest.java [hunk 7]
```diff
@@ -167,20 +167,36 @@ public void testIterateOnMapOfLists() throws Exception {
 
   @Test
   public void testParseMap() {
-    assertEquals(Utils.parseCommaSeparatedStringMapFromString("", "test_field").size(), 0);
-    Map nonEmptyMap = Utils.parseCommaSeparatedStringMapFromString("a=b", "test_field");
+    assertEquals(Utils.parseJsonMapFromString("", "test_field").size(), 0);
+    Map nonEmptyMap = Utils.parseJsonMapFromString("a=b", "test_field");
     Map expectedMap = new HashMap<>();
     expectedMap.put("a", "b");
     assertEquals(nonEmptyMap, expectedMap);
```
Comment (nisargthakkar) [inline internal/venice-common/src/test/java/com/linkedin/venice/utils/UtilsTest.java:174]: This test case should now be failing. Let's change this test to ensure that an exception is thrown for comma-separated map similar to the `expectThrows` in the next line

File: internal/venice-common/src/test/java/com/linkedin/venice/utils/UtilsTest.java [hunk 8]
```diff
@@ -167,20 +167,36 @@ public void testIterateOnMapOfLists() throws Exception {
...
 
-    VeniceException e = expectThrows(
-        VeniceException.class,
-        () -> Utils.parseCommaSeparatedStringMapFromString("invalid_value", "test_field"));
+    VeniceException e =
+        expectThrows(VeniceException.class, () -> Utils.parseJsonMapFromString("invalid_value", "test_field"));
     assertTrue(e.getMessage().contains("must be key value pairs separated by comma"));
```
Comment (nisargthakkar) [inline internal/venice-common/src/test/java/com/linkedin/venice/utils/UtilsTest.java:178]: This error message should also be updated

File: internal/venice-common/src/test/java/com/linkedin/venice/utils/UtilsTest.java [hunk 9]
```diff
@@ -167,20 +167,36 @@ public void testIterateOnMapOfLists() throws Exception {
...
+  public void testParseJsonMapFromString() {
+    try {
+      Utils.parseJsonMapFromString(
+          "{\"changeCaptureView\": {\"viewClassName\": \"com.linkedin.venice.views.ChangeCaptureView\",\"params\": {}}}",
+          "someField"); // Method that should not throw an exception
+    } catch (Exception e) {
+      fail("Expected no exception to be thrown, but got: " + e.getMessage());
+    }
+
+    Assert.assertThrows(
```
Comment (nisargthakkar) [inline internal/venice-common/src/test/java/com/linkedin/venice/utils/UtilsTest.java:196]: Can you add a comment about why this throws an exception? I guess it is because the string is not a valid JSON, right?

File: internal/venice-common/src/main/java/com/linkedin/venice/utils/Utils.java (hunk 4 above)
Comment (nisargthakkar) [reply internal/venice-common/src/main/java/com/linkedin/venice/utils/Utils.java]: Thanks. I left some more comments. But it is looking quite good now

Comment (elijahgrimaldi) [reply internal/venice-common/src/main/java/com/linkedin/venice/utils/Utils.java]: I realized the testParseMap test was covering this specific function, so I removed my redundant test and updated that test to check for the new functionality and added the other changes. Let me know if this is fine.

File: internal/venice-common/src/test/java/com/linkedin/venice/utils/UtilsTest.java [hunk 10]
```diff
@@ -1,9 +1,6 @@
 package com.linkedin.venice.utils;
 
-import static org.testng.Assert.assertEquals;
-import static org.testng.Assert.assertTrue;
-import static org.testng.Assert.expectThrows;
-import static org.testng.Assert.fail;
+import static org.testng.Assert.*;
```
Comment (nisargthakkar) [inline internal/venice-common/src/test/java/com/linkedin/venice/utils/UtilsTest.java:3]: Can you please remove wildcard imports? If you're using intellij, try this: https://blog.marcnuri.com/intellij-idea-how-to-disable-wildcard-imports

The rest LGTM

**Review by nisargthakkar — COMMENTED** [review]
Can you please take a look at the following tests:
```
com.linkedin.venice.TestAdminTool > testAdminUpdateStoreArg
com.linkedin.venice.TestAdminTool > testAdminConfigureView
```

File: internal/venice-common/src/main/java/com/linkedin/venice/utils/Utils.java (hunk 4 above)
Comment (nisargthakkar) [reply internal/venice-common/src/main/java/com/linkedin/venice/utils/Utils.java]: Yeah, that works too. Thanks. The change mostly looks good. There are a few more failing tests. Can you please address those too?

**Review by nisargthakkar — APPROVED** [review]
LGTM

Thanks a lot for the contribution @elijahgrimaldi. Great job.

I have a suggestion for future PRs. Can you please try to keep the pull request template? For this PR, I have updated the PR description to demonstrate the format that we expect the PR description to follow.

**Comment by elijahgrimaldi** [comment]
> LGTM
> 
> Thanks a lot for the contribution @elijahgrimaldi. Great job.
> 
> I have a suggestion for future PRs. Can you please try to keep the pull request template? For this PR, I have updated the PR description to demonstrate the format that we expect the PR description to follow.
Yes sorry about that. I will make sure to do it next time.
